pkg/bootstrap/static.go

```diff
@@ -0,0 +1,392 @@
+		},
+	}
+
+	if opts.ZipkinAddress != "" {
+		zipkinAddress, err := toAddress(opts.ZipkinAddress)
+		if err != nil {
+			return nil, err
+		}
+		out.StaticResources.Clusters = append(out.StaticResources.Clusters,
+			v2.Cluster{
+				Name:           ZipkinCluster,
+				ConnectTimeout: 1 * time.Second,
+				Type:           v2.Cluster_STRICT_DNS,
+				LbPolicy:       v2.Cluster_ROUND_ROBIN,
+				Hosts:          []*core.Address{zipkinAddress},
+			})
+		out.Tracing = &trace.Tracing{
+			Http: &trace.Tracing_Http{
+				Name: util.Zipkin,
+				Config: toStruct(&trace.ZipkinConfig{
+					CollectorCluster:  ZipkinCluster,
+					CollectorEndpoint: "/api/v1/spans",
+				}),
+			},
+		}
+	}
+
+	// telemetry bootstrap should route telemetry directly to the local upstream
+	telemetryCluster := TelemetryCluster
+	if opts.TelemetryUpstream == nil {
+		telemetryAddress, err := toAddress(opts.TelemetryAddress)
+		if err != nil {
+			return nil, err
+		}
+		cluster := v2.Cluster{
+			Name:                 TelemetryCluster,
+			ConnectTimeout:       1 * time.Second,
+			Type:                 v2.Cluster_STRICT_DNS,
+			LbPolicy:             v2.Cluster_ROUND_ROBIN,
+			Hosts:                []*core.Address{telemetryAddress},
+			CircuitBreakers:      circuitBreakers,
+			Http2ProtocolOptions: &core.Http2ProtocolOptions{},
+		}
+		if opts.ControlPlaneAuth {
+			cluster.TlsContext = &auth.UpstreamTlsContext{
+				CommonTlsContext: commonTLSContext([]string{opts.TelemetrySAN}),
+			}
+		}
+		out.StaticResources.Clusters = append(out.StaticResources.Clusters, cluster)
+	} else {
+		telemetryCluster = clusterName(*opts.TelemetryUpstream)
+	}
+
+	out.StaticResources.Listeners = BuildListeners(opts.Upstreams, telemetryCluster, opts.ControlPlaneAuth)
+
+	if err := out.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &out, nil
+}
+
+func commonTLSContext(altNames []string) *auth.CommonTlsContext {
+	return &auth.CommonTlsContext{
+		TlsCertificates: []*auth.TlsCertificate{{
+			CertificateChain: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/cert-chain.pem"},
+			},
+			PrivateKey: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/key.pem"},
+			},
+		}},
+		ValidationContext: &auth.CertificateValidationContext{
+			TrustedCa: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/root-cert.pem"},
+			},
+			VerifySubjectAltName: altNames,
+		},
+	}
+}
+
+func toAddress(address string) (*core.Address, error) {
+	host, portStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return nil, err
+	}
+	return &core.Address{
+		Address: &core.Address_SocketAddress{
+			SocketAddress: &core.SocketAddress{
+				Address:       host,
+				PortSpecifier: &core.SocketAddress_PortValue{PortValue: uint32(port)},
+			},
+		},
+	}, nil
+}
+
+func toStruct(msg proto.Message) *types.Struct {
```

Comment:
This also already exists: https://github.com/istio/istio/blob/97882cede142f5ac246a2e7cff4d6ff6bf631389/pilot/pkg/networking/util/util.go#L163

Reply:
I don't want to import pilot into this top-level `pkg`

Reply:
Then move this function up to `pkg/proto/struct.go` and use it from there.

Reply:
We can consolidate the code once the dust settles.

Reply:
No, duplicating this type of basic function all over the place is debt there is literally zero reason to incur. Just move the function up to the top level, update the two places that use it, and call it a day. We're not in such a state of urgency that simple code health needs to be skipped.

Reply:
I agree with Zack here. Its one thing to programmatically generate the config, but its another to literally duplicate all the code. There will be no single authoritative place to clean things up.

If all you want is a bunch of generic functions to build clusters, add filters, etc. refactor the code in networking/core/v1alpha3/listeners|clusters etc., into smaller packages, like listeners/build.go, util.go .. and just reuse those utility functions.

Reply:
I don't want to refactor pilot code, and I don't see how this is a blocker for this PR. Feel free to refactor and combine/unify these two pieces later. But this is just step 1 to get telemetry out, most of the work is happening in mixer/dashboards, and is currently blocked by lack of telemetry.

Reply:
In particular, what does v1alpha3 have to do with the bootstraps? Why do we want to depend on some networking API version for a static bootstrap?